src/resmom/linux/mom_mach.c

```diff
@@ -8001,6 +8008,63 @@ pidcache_create(void)
 	return (pidcache_arena);
 }
 
+/** @fn pidcache_reset
+ * @brief	reset pidcache with new set of PIDs from /dev/cpuset/PBSPro
+ *
+ * @return	int
+ * @retval	nprocs	- number of processes cached
+ *		-1	- when stat fails to open a cpuset tasks file
+ *
+ * @par MT-Safe:	yes
+ * @par Side Effects:
+ *	None
+ */
+static int
+pidcache_reset(pidcachetype_t *pidcache)
+{
+	struct dirent		*dent;
+	pid_t			pid;
+	char			taskname[256];
+	int			ncantstat = 0;
+	FILE			*fd;
+	char			line[25] = {0};
+	char			*p;
+	int			i;
+	int			nprocs = 0;
+
+	if (cpusetdir == NULL) {
+		if ((cpusetdir = opendir(cpusetfs)) == NULL) {
+			log_err(errno, __func__, "opendir");
+			return -1;
+		}
+	} else
+		rewinddir(cpusetdir);
+
+	while (errno = 0, (dent = readdir(cpusetdir)) != NULL) {
+		if (!isdigit(dent->d_name[0]))
+			continue;
+
+		sprintf(taskname, "%s/%s/tasks", cpusetfs, dent->d_name);
+
+		if ((fd = fopen(taskname, "r")) == NULL) {
+			ncantstat++;
+			continue;
+		}
+
+		for (i=0; (p = fgets(line, sizeof(line), fd)) != NULL; i++) {
+			pid = strtol(p, NULL, 10);
+			if (pidcache_insert(pid, pidcache))
+				nprocs++;
+		}
+		fclose(fd);
+	}
+
+	if (ncantstat)
+		return -1;
+	else
+		return nprocs;
```

Comment:
I believe we should always return nprocs if (nprocs > 0). The return value is used to log the number of pids that are cached. 
We should return -1 only if we do not add any pid to the cache...
With the current logic even if we have added pids to the cache, we would return -1, and report ncached to be zero...